napi/src/threadsafe_function.rs

```diff
@@ -0,0 +1,211 @@
+use crate::{check_status, ptr, sys, Env, Function, Result, Value};
+use std::os::raw::{c_char, c_void};
+
+use sys::napi_threadsafe_function_call_mode;
+use sys::napi_threadsafe_function_release_mode;
+
+pub trait ToJs: Copy + Clone {
+  type Output;
+  type JsValue;
+
+  fn resolve(
+    &self,
+    env: &mut Env,
+    output: &mut Self::Output,
+  ) -> Result<(u64, Value<Self::JsValue>)>;
+}
+
+/// An example of using `ThreadsafeFunction`:
+///
+/// ```
+/// // Define a struct for handling the data passed from `ThreadsafeFunction::call`
+/// // and return the data to be used for the js callback.
+/// #[derive(Clone, Copy)]
+/// struct HandleNumber;
+///
+/// impl ToJs for HandleNumber {
+///   type Output = u8;
+///   type JsValue = Number;
+///
+///   fn resolve(&self, env: &mut Env, output: &mut Self::Output) -> Result<(u64, Value<Self::JsValue>)> {
+///     let argv: u64 = 1;
+///     let value = env.create_uint32(*output as u32)?;
+///     Ok((argv, value))
+///   }
+/// }
+///
+///  #[js_function(1)]
+///  fn test_threadsafe_function(ctx: CallContext) -> Result<Value<Undefined>> {
+///    // The callback function from js which will be called in `ThreadsafeFunction::call`.
+///    let func: Value<Function> = ctx.get::<Function>(0)?;
+///
+///    let to_js = HandleNumber;
+///    let tsfn = ThreadsafeFunction::create(*ctx.env, func, to_js, 0)?;
+///
+///    thread::spawn(move || {
+///      let output: u8 = 42;
+///      // It's okay to call a threadsafe function multiple times.
+///      tsfn.call(Ok(output), napi_tsfn_blocking).unwrap();
+///      tsfn.call(Ok(output), napi_tsfn_blocking).unwrap();
+///      // We should call `ThreadsafeFunction::release` manually when we don't
+///      // need the instance anymore, or it will prevent Node.js from exiting
+///      // automatically and possiblely cause memory leaks.
+///      tsfn.release(napi_tsfn_release).unwrap();
+///    });
+///
+///    Ok(Env::get_undefined(ctx.env)?)
+///  }
+/// ```
+#[derive(Debug, Clone, Copy)]
+pub struct ThreadsafeFunction<T: ToJs> {
+  raw_value: sys::napi_threadsafe_function,
+  to_js: T,
+}
+
+unsafe impl<T: ToJs> Send for ThreadsafeFunction<T> {}
+unsafe impl<T: ToJs> Sync for ThreadsafeFunction<T> {}
+
+impl<T: ToJs> ThreadsafeFunction<T> {
+  /// See [napi_create_threadsafe_function](https://nodejs.org/api/n-api.html#n_api_napi_create_threadsafe_function)
+  /// for more information.
+  pub fn create(env: Env, func: Value<Function>, to_js: T, max_queue_size: u64) -> Result<Self> {
```

Comment:
`env: &Env` is okay, so you needn't call `*ctx.env` in `js_function`

Reply:
I have also changed the `refer` and `unref` to use a reference of `Env`.